jerry-core/api/jerry.c

```diff
@@ -2260,37 +2260,48 @@ jerry_create_instance (uint32_t heap_size, /**< the size of heap */
 {
 #ifdef JERRY_ENABLE_EXTERNAL_CONTEXT
 
-  uint32_t lcache_size = 0;
+#ifndef JERRY_SYSTEM_ALLOCATOR
+  /* Minimum heap size is 1Kbyte. */
+  if (heap_size < 1024)
+  {
+    return NULL;
+  }
+#else
+  heap_size = 0;
+#endif /* !JERRY_SYSTEM_ALLOCATOR */
+
+  size_t total_size = sizeof (jerry_instance_t) + heap_size + JMEM_ALIGNMENT - 1;
 
 #ifndef CONFIG_ECMA_LCACHE_DISABLE
-  lcache_size = sizeof (jerry_hash_table_t);
+  total_size += sizeof (jerry_hash_table_t) + JMEM_ALIGNMENT - 1;
 #endif /* !CONFIG_ECMA_LCACHE_DISABLE */
 
-  uint32_t alloc_size = (uint32_t) (sizeof (jerry_instance_t) + lcache_size);
-
-#ifndef JERRY_SYSTEM_ALLOCATOR
-  alloc_size = alloc_size + heap_size + JMEM_ALIGNMENT - 1;
-#endif /* !JERRY_SYSTEM_ALLOCATOR */
+  total_size = JERRY_ALIGNUP (total_size, JMEM_ALIGNMENT);
 
-  jerry_instance_t *instance_p = (jerry_instance_t *) alloc (alloc_size, cb_data_p);
-  memset (instance_p, 0, alloc_size);
+  jerry_instance_t *instance_p = (jerry_instance_t *) alloc (total_size, cb_data_p);
 
   if (instance_p == NULL)
   {
     return NULL;
   }
 
-  instance_p->heap_size = heap_size;
-  instance_p->lcache_p = instance_p->buffer;
+  memset (instance_p, 0, total_size);
+
+  uint8_t *byte_p = (uint8_t *) instance_p;
+  byte_p += JERRY_ALIGNUP (sizeof (jerry_instance_t), JMEM_ALIGNMENT);
 
 #ifndef JERRY_SYSTEM_ALLOCATOR
-  uint8_t *unaligned_heap_p = instance_p->lcache_p + lcache_size;
-  uintptr_t alignment_mask = ~(uintptr_t) (JMEM_ALIGNMENT - 1);
-  instance_p->heap_p = (uint8_t *) (((uintptr_t) unaligned_heap_p + JMEM_ALIGNMENT - 1) & alignment_mask);
+  instance_p->heap_p = (jmem_heap_t *) byte_p;
+  instance_p->heap_size = heap_size;
```

Comment:
I don't think it solve the alignment problem.

if instance_p is not multiple of 8bytes, `JERRY_ALIGNUP (sizeof (jerry_instance_t), JMEM_ALIGNMENT)` is multiple of 8bytes, then `heap_p = instance_p + JERRY_ALIGNUP (sizeof (jerry_instance_t), JMEM_ALIGNMENT)` is not multiple of 8 bytes